tempesta_fw/sock_srv.c

```diff
@@ -946,6 +959,90 @@ tfw_cfgop_lookup_sg_cfg(const char *name)
 	return NULL;
 }
 
+static int
+tfw_cfgop_srv_create_health(const char *hname, TfwServer *new_srv,
+			    TfwServer *orig_srv)
+{
+	TfwSrvHealth *health;
+	TfwServer *srv;
+	size_t size = 0;
+	bool orig_hm = false;
+
+	if (orig_srv)
+		orig_hm = test_bit(TFW_SRV_B_HMONITOR,
+				   (unsigned long *)&orig_srv->hm_flags);
+
+	/*
+	 * Nothing to do if the same server with the same
+	 * hmonitor, or the same server without hmonitor (and
+	 * new hmonitor is not specified), or new server
+	 * without hmonitor at all.
+	 */
+	if ((hname && orig_hm && tfw_apm_hm_srv_eq(hname, orig_srv))
+	    || (!hname && orig_srv && !orig_hm)
+	    || (!hname && !orig_srv))
+		return 0;
+
+	if (hname)
+		size = strlen(hname) + 1;
+	health = kzalloc(sizeof(TfwSrvHealth) + size, GFP_KERNEL);
+	if (!health) {
+		TFW_ERR_NL("can't allocate memory for server health entry\n");
+		return -ENOMEM;
+	}
+
+	if (hname) {
+		if (orig_hm)
+			health->flags |= TFW_SRV_HEALTH_DISABLE;
+		health->flags |= TFW_SRV_HEALTH_ENABLE;
+		health->name = (char *)(health + 1);
+		memcpy(health->name, hname, size);
+		srv = orig_srv ? : new_srv;
+	} else {
+		srv = orig_srv;
```

Comment:
It's difficult to understand the variety of combinations of function's arguments without comments. `orig_srv` may be `NULL`. The code suggest it's definitely NOT `NULL` when `!hname`. Why is that? A proper comment would be helpful.

Reply:
This has been redesigned.